### PR TITLE
[MLA-488] Fix observations in demonstration drawer

### PR DESCRIPTION
**Pull request by chriselion**
### Proposed change(s)

Fix a long-standing bug in the DemonstrationDrawer. It was never updated to handle the changes to Observations.

PushBlock demo, which has multiple 1-D sensors:
image

GridWorld demo, which has 1 multi-D sensor:
image

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)

https://jira.unity3d.com/browse/MLA-488

### Types of change(s)

- [x] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs [hunk 1]
```diff
@@ -80,6 +99,25 @@ public static BrainParametersProto ToProto(this BrainParameters bp, string name,
             return brainParametersProto;
         }
 
+        /// <summary>
+        /// Convert a BrainParametersProto to a BrainParameters struct.
+        /// </summary>
+        /// <param name="bpp">An instance of a brain parameters protobuf object.</param>
+        /// <returns>A BrainParameters struct.</returns>
+        public static BrainParameters ToBrainParameters(this BrainParametersProto bpp)
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs:107]: Just moved this closer to the BrainParameters -> BrainParametersProto conversion

File: com.unity.ml-agents/Runtime/Demonstrations/Demonstration.cs [hunk 2]
```diff
@@ -1,38 +0,0 @@
...
-using MLAgents.Policies;
-
-namespace MLAgents.Demonstrations
-{
-    /// <summary>
-    /// Demonstration Object. Contains meta-data regarding demonstration.
-    /// Used for imitation learning, or other forms of learning from data.
-    /// </summary>
-    [Serializable]
-    internal class Demonstration : ScriptableObject
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Demonstrations/Demonstration.cs:12]: Split this file into DemonstrationSummary.cs and DemonstrationMetaData.cs

**Comment by chriselion** [comment]
Old inspector for comparison: 
image

**Comment by chriselion** [comment]
Note: need to update the inspector screenshot here https://github.com/Unity-Technologies/ml-agents/blob/master/docs/images/demo_inspector.png - I'll do that after we've agreed on the layout format.

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationSummary.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace MLAgents.Demonstrations
+{
+    /// <summary>
+    /// Summary of a loaded Demonstration file. Only used for display in the Inspector.
+    /// </summary>
+    [Serializable]
+    internal class DemonstrationSummary : ScriptableObject
+    {
+        public DemonstrationMetaData metaData;
```
Comment (vincentpierre) [inline com.unity.ml-agents/Runtime/Demonstrations/DemonstrationSummary.cs:14]: Does the DemonstrationSummary need DemonstrationMetaData ? I thought the two were side by side anyway.

File: Project/Assets/ML-Agents/Examples/Bouncer/Demos/ExpertBouncer.demo.meta [hunk 4]
```diff
@@ -4,7 +4,7 @@ ScriptedImporter:
   fileIDToRecycleName:
     11400000: Assets/ML-Agents/Examples/Bouncer/Demos/ExpertBouncer.demo
   externalObjects: {}
-  userData: ' (MLAgents.Demonstrations.Demonstration)'
+  userData: ' (MLAgents.Demonstrations.DemonstrationSummary)'
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Bouncer/Demos/ExpertBouncer.demo.meta:7]: Does this mean that old demo files will be incompatible (nothing wrong with that, just making sure)

File: com.unity.ml-agents/Editor/DemonstrationImporter.cs [hunk 5]
```diff
@@ -36,17 +37,30 @@ public override void OnImportAsset(AssetImportContext ctx)
...
+                }
+
                 reader.Close();
 
-                var demonstration = ScriptableObject.CreateInstance<Demonstration>();
-                demonstration.Initialize(brainParameters, metaData);
-                userData = demonstration.ToString();
+                var demonstrationSummary = ScriptableObject.CreateInstance<DemonstrationSummary>();
+                demonstrationSummary.Initialize(brainParameters, metaData, observationSummaries);
+                userData = demonstrationSummary.ToString();
```
Comment (vincentpierre) [inline com.unity.ml-agents/Editor/DemonstrationImporter.cs:57]: Where is this variable used ?

File: com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs [hunk 6]
```diff
@@ -59,6 +60,24 @@ public static AgentInfoProto ToAgentInfoProto(this AgentInfo ai)
             return agentInfoProto;
         }
 
+        public static List<ObservationSummary> GetObservationSummaries(this AgentInfoActionPairProto infoActionPair)
```
Comment (vincentpierre) [inline com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs:63]: The method name is very explicit but maybe a method comment?

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationSummary.cs (hunk 3 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Demonstrations/DemonstrationSummary.cs:14]: The fields from the DemonstrationMetaData are displayed in the inspector. DemonstrationSummary is _only_ used by the DemonstrationDrawer, it's not a full Demonstration (that's why I renamed it).

File: com.unity.ml-agents/Editor/DemonstrationImporter.cs (hunk 5 above)
Comment (chriselion) [reply com.unity.ml-agents/Editor/DemonstrationImporter.cs:57]: I'm not really sure about the purpose of it. It sets this field https://docs.unity3d.com/ScriptReference/AssetImporter-userData.html which explains the .meta file changes.

File: Project/Assets/ML-Agents/Examples/Bouncer/Demos/ExpertBouncer.demo.meta (hunk 4 above)
Comment (chriselion) [reply Project/Assets/ML-Agents/Examples/Bouncer/Demos/ExpertBouncer.demo.meta:7]: No, shouldn't have any effect.

File: com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs (hunk 6 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs:68]: Updated.

**Pull request merged**